lib/msf/core/modules/external/go/src/metasploit/module/core.go

```diff
@@ -0,0 +1,156 @@
+/*
+ * Defines core functionality for a GOLANG module
+ */
+
+package module
+
+import (
+	"bufio"
+	"encoding/json"
+	"log"
+	"os"
+	"strings"
+	"errors"
+)
+
+/*
+ * RunCallback represents the method to call from the module
+ */
+type RunCallback func(params map[string]interface{})
+
+/*
+ * Initializes the module waiting for input from stdin
+ */
+func Init(metadata *Metadata, callback RunCallback) {
+	var req Request
+
+	err := json.NewDecoder(os.Stdin).Decode(&req)
+	if err != nil {
+		log.Fatalf("could not decode JSON: %v", err)
+	}
+
+	switch strings.ToLower(req.Method) {
+	case "describe":
+		metadata.Capabilities = []string{"run"}
+		res := &MetadataResponse{"2.0", req.ID, metadata}
+		if err := rpcSend(res); err != nil {
+			log.Fatalf("error on running %s: %v", req.Method, err)
+		}
+	case "run":
+		params, e := parseParams(req.Parameters)
+		if e != nil {
+			log.Fatal(e)
+		}
+		callback(params)
+		res := &RunResponse{"2.0", req.ID, RunResult{"Module complete", ""}}
+		if err := rpcSend(res); err != nil {
+			log.Fatalf("error on running %s: %v", req.Method, err)
+		}
+	default:
+		log.Fatalf("method %s not implemented yet", req.Method)
+	}
+}
+
+type (
+	Request struct {
+		Jsonrpc string `json:"jsonrpc"`
+		Method  string `json:"method"`
+		ID      string `json:"id"`
+		Parameters interface{} `json:"params"`
+	}
+
+	MetadataResponse struct {
+		Jsonrpc string    `json:"jsonrpc"`
+		ID      string    `json:"id"`
+		Result  *Metadata `json:"result"`
+	}
+
+	RunResult struct {
+		Message string `json:"message"`
+		Return  string `json:"return"`
+	}
+
+	RunResponse struct {
+		Jsonrpc string    `json:"jsonrpc"`
+		ID      string    `json:"id"`
+		Result  RunResult `json:"result"`
+	}
+
+	Parameters struct {
+		Type string `json:"type"`
+		Description string `json:"description"`
+		Required bool `json:"required"`
+		Default interface{} `json:"default"`
+	}
+)
+
+type response struct {
+	Jsonrpc string `json:"jsonrpc"`
+	ID      string `json:"id"`
+}
+
+func parseParams(passedParams interface{}) (map[string]interface{}, error) {
+	v, ok := passedParams.(map[string]interface{})
+	if !ok {
+		return nil, errors.New("cannot parse values")
+	}
+
+	return v, nil
+}
+
+func rpcSend(res interface{}) error {
+	resStr, err := json.Marshal(res)
+	if err != nil {
+		return err
+	}
+	f := bufio.NewWriter(os.Stdout)
+	if _, err := f.Write(resStr); err != nil {
+		return err
+	}
+	if err := f.Flush(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type (
+	logparams struct {
+		Level   string `json:"level"`
+		Message string `json:"message"`
+	}
+
+	logRequest struct {
+		Jsonrpc string    `json:"jsonrpc"`
+		Method  string    `json:"method"`
+		Params  logparams `json:"params"`
+	}
+)
+
+func LogInfo(message string) {
+	msfLog(message, "info")
+}
+
+func LogError(message string) {
+	msfLog(message, "error")
+}
+
+func msfLog(message string, level string) {
```

Comment:
This method should behave more like the `log` method in `lib/msf/core/modules/external/python/metasploit/cli.py`, otherwise, all callers become responsible for adding the leading log status indicators (`[*]`, `[!]`, `[+]`, etc) to the actual messages themselves. This could result in a lot of inconsistencies.

Reply:
I think the 'level' string makes metasploit display the correct character at the beginning of the message already.

Reply:
cli.py is something that interprets the output of the message that comes _from_ a module like this, so it's a different thing.

Reply:
Yep, you are right.

Reply:
also, all of the 'mux.Lock/Unlock' code needs to be fixed